File: scripts/importar.rb
#!/usr/bin/env ruby
# scripts/importar.rb
# Importa conteúdo do WordPress REST API para as coleções Jekyll.
# Uso: ruby scripts/importar.rb
# Requer apenas stdlib Ruby (net/http, json, yaml, fileutils, cgi, set).

require 'net/http'
require 'net/https'
require 'json'
require 'yaml'
require 'fileutils'
require 'cgi'
require 'set'
require 'uri'

API_BASE    = ENV['WP_API_BASE'] || 'https://gabrielcanowp-djfpn.wpcomstaging.com/wp-json/wp/v2'
ROOT_DIR    = File.expand_path('..', __dir__)
CARROS_DIR  = ENV['CARROS_DIR']  || File.join(ROOT_DIR, '_carros')
BANNERS_DIR = ENV['BANNERS_DIR'] || File.join(ROOT_DIR, '_banners')

# ---------------------------------------------------------------------------
# HTTP helpers
# ---------------------------------------------------------------------------

def wp_get(path, params = {})
  uri = URI("#{API_BASE}#{path}")
  uri.query = URI.encode_www_form(params) unless params.empty?

  http = Net::HTTP.new(uri.host, uri.port)
  http.use_ssl = (uri.scheme == 'https')
  http.read_timeout = 30
  http.open_timeout = 10

  request = Net::HTTP::Get.new(uri)
  request['User-Agent'] = 'Jekyll-Importer/1.0'
  request['Accept']     = 'application/json'

  response = http.request(request)

  unless response.is_a?(Net::HTTPSuccess)
    raise "HTTP #{response.code} para #{uri}"
  end

  body        = JSON.parse(response.body)
  total_pages = response['X-WP-TotalPages']&.to_i || 1

  [body, total_pages]
end

def wp_get_all(path, params = {})
  page       = 1
  all_items  = []

  loop do
    items, total_pages = wp_get(path, params.merge(page: page, per_page: 100))

    # Quando não há itens o WP retorna [] (array vazio)
    break if items.is_a?(Array) && items.empty?

    all_items.concat(items)
    break if page >= total_pages

    page += 1
  end

  all_items
rescue => e
  warn "  AVISO: falha ao buscar #{path} — #{e.message}"
  []
end

# ---------------------------------------------------------------------------
# Taxonomy / embed helpers
# ---------------------------------------------------------------------------

def extract_terms(embedded, taxonomy)
  term_groups = embedded&.dig('wp:term') || []
  term_groups.flatten
             .select  { |t| t.is_a?(Hash) && t['taxonomy'] == taxonomy }
             .map     { |t| CGI.unescapeHTML(t['name'].to_s).strip }
             .reject  { |n| n.empty? }
end

def first_term(embedded, taxonomy)
  extract_terms(embedded, taxonomy).first
end

def featured_url(embedded)
  media = embedded&.dig('wp:featuredmedia')
  return nil unless media.is_a?(Array) && !media.empty?

  first = media.first
  return nil unless first.is_a?(Hash)

  first['source_url']
end

# ---------------------------------------------------------------------------
# Gallery
# ---------------------------------------------------------------------------

def fetch_gallery(post_id, featured)
  media_items, = wp_get('/media', parent: post_id, per_page: 100)
  urls = media_items.map { |m| m['source_url'] }.compact.uniq

  if featured
    urls.delete(featured)
    urls.unshift(featured)
  end

  urls
rescue => e
  warn "  AVISO: galeria do post #{post_id} — #{e.message}"
  featured ? [featured] : []
end

# ---------------------------------------------------------------------------
# File helpers
# ---------------------------------------------------------------------------

def clean_wp_files(dir)
  FileUtils.mkdir_p(dir)

  Dir.glob(File.join(dir, '*.md')).each do |path|
    content = File.read(path)
    if content =~ /^wp_id:/
      File.delete(path)
      puts "    Removido: #{File.basename(path)}"
    else
      puts "    Mantido (sem wp_id): #{File.basename(path)}"
    end
  end
end

# Serializa um valor escalar string como YAML seguro.
# Usa aspas duplas quando o valor contém caracteres especiais.
YAML_SPECIAL = /[:#\{\}\[\],|>&*?!%@`"'\\]|\A[-\s]|\s\z|\A(true|false|null|~|\d+)\z/i.freeze

def yaml_scalar(val)
  return '""' if val.nil? || val.empty?

  if val =~ YAML_SPECIAL || val.include?("\n")
    val.inspect  # gera "string com escaping Ruby" — válido em YAML
  else
    val
  end
end

# Serializa o hash de frontmatter em YAML com bloco de início/fim "---"
# Garante que campos nulos apareçam como chave vazia (ex: "km:\n")
def render_frontmatter(hash)
  lines = ["---"]

  hash.each do |key, val|
    case val
    when nil
      lines << "#{key}:"
    when Array
      if val.empty?
        lines << "#{key}: []"
      else
        lines << "#{key}:"
        val.each { |item| lines << "  - #{yaml_scalar(item.to_s)}" }
      end
    when String
      lines << "#{key}: #{yaml_scalar(val)}"
    else
      lines << "#{key}: #{val}"
    end
  end

  lines << "---"
  lines.join("\n") + "\n"
end

def write_md(path, frontmatter, body = nil)
  content = render_frontmatter(frontmatter)
  content += "\n#{body.strip}\n" if body && !body.strip.empty?
  File.write(path, content)
end

# ---------------------------------------------------------------------------
# Import: Carros
# ---------------------------------------------------------------------------

def import_carros
  puts "\n==> Importando carros..."

  carros        = wp_get_all('/carro', _embed: 1)
  total         = carros.size
  imported      = 0
  seen_slugs    = Set.new

  puts "    #{total} carro(s) encontrado(s) na API"

  carros.each_with_index do |carro, idx|
    raw_slug = carro['slug'].to_s.strip
    raw_slug = "carro-#{carro['id']}" if raw_slug.empty?

    # Detectar slug duplicado
    slug = raw_slug
    if seen_slugs.include?(slug)
      slug = "#{raw_slug}-#{carro['id']}"
      warn "  AVISO: slug duplicado '#{raw_slug}', renomeado para '#{slug}'"
    end
    seen_slugs << slug

    embedded = carro['_embedded'] || {}

    # Taxonomias
    marca      = first_term(embedded, 'marca')
    modelo     = first_term(embedded, 'modelo')
    cambio     = first_term(embedded, 'cambio')
    combustivel = first_term(embedded, 'combustivel')
    cor        = first_term(embedded, 'cor')
    opcionais  = extract_terms(embedded, 'opcional')

    anos = extract_terms(embedded, 'ano').sort
    ano_str = anos.empty? ? nil : anos.join('/')

    # Título do WordPress (prioridade)
    raw_title = CGI.unescapeHTML(carro.dig('title', 'rendered').to_s).strip
    
    # Título composto (fallback se rendered estiver vazio)
    composed  = [marca, modelo, ano_str].compact.join(' ')
    title     = raw_title.empty? ? (composed.empty? ? slug : composed) : raw_title

    # Imagens
    featured   = featured_url(embedded)
    imagens    = fetch_gallery(carro['id'], featured)

    # Descrição (HTML do WP, usado como body do markdown)
    body = CGI.unescapeHTML(carro.dig('content', 'rendered').to_s).strip

    frontmatter = {
      'wp_id'       => carro['id'],
      'title'       => title.empty? ? slug : title,
      'marca'       => marca,
      'modelo'      => modelo,
      'ano'         => ano_str,
      'km'          => nil,
      'preco'       => nil,
      'cambio'      => cambio,
      'combustivel' => combustivel,
      'cor'         => cor,
      'portas'      => nil,
      'destaque'    => nil,
      'opcionais'   => opcionais,
      'imagens'     => imagens,
    }

    filepath = File.join(CARROS_DIR, "#{slug}.md")
    write_md(filepath, frontmatter, body)

    imported += 1
    puts "  [#{idx + 1}/#{total}] #{slug}"
  rescue => e
    warn "  ERRO ao importar carro ID #{carro['id']}: #{e.message}"
  end

  imported
end

# ---------------------------------------------------------------------------
# Import: Banners
# ---------------------------------------------------------------------------

def import_banners
  puts "\n==> Importando banners..."

  banners  = wp_get_all('/banner', _embed: 1)
  total    = banners.size
  imported = 0

  puts "    #{total} banner(s) encontrado(s) na API"

  banners.each_with_index do |banner, idx|
    slug = banner['slug'].to_s.strip
    slug = "banner-#{banner['id']}" if slug.empty?

    embedded  = banner['_embedded'] || {}
    imagem    = featured_url(embedded)
    title     = CGI.unescapeHTML(banner.dig('title', 'rendered').to_s).strip

    frontmatter = {
      'wp_id'     => banner['id'],
      'title'     => title.empty? ? slug : title,
      'imagem'    => imagem,
      'link'      => nil,
      'cta'       => nil,
      'subtitulo' => nil,
      'ativo'     => true,
      'ordem'     => idx + 1,
    }

    filepath = File.join(BANNERS_DIR, "#{slug}.md")
    write_md(filepath, frontmatter)

    imported += 1
    puts "  [#{idx + 1}/#{total}] #{slug}"
  rescue => e
    warn "  ERRO ao importar banner ID #{banner['id']}: #{e.message}"
  end

  imported
end

# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def run_import
  FileUtils.mkdir_p(CARROS_DIR)
  FileUtils.mkdir_p(BANNERS_DIR)

  puts '==> Limpando arquivos gerados anteriormente...'
  puts "    _carros/:"
  clean_wp_files(CARROS_DIR)
  puts "    _banners/:"
  clean_wp_files(BANNERS_DIR)

  carros_importados  = import_carros
  banners_importados = import_banners

  puts "\n==> Concluído!"
  puts "    Carros importados : #{carros_importados}"
  puts "    Banners importados: #{banners_importados}"
  
  {
    carros: carros_importados,
    banners: banners_importados
  }
end

if __FILE__ == $0
  run_import
  puts "\nPróximo passo: bundle exec jekyll serve"
end

File: scripts/importar.py
#!/usr/bin/env python3
import os
import re
import json
import yaml
import requests
import html
from urllib.parse import urlencode
from pathlib import Path
from dotenv import load_dotenv

load_dotenv(Path(__file__).parent.parent / '.env')

# Configurações
API_BASE = os.getenv('WP_API_BASE', 'https://gabrielcanowp-djfpn.wpcomstaging.com/wp-json/wp/v2')
ROOT_DIR = Path(__file__).parent.parent.absolute()
CARROS_DIR = os.getenv('CARROS_DIR', str(ROOT_DIR / '_carros'))
BANNERS_DIR = os.getenv('BANNERS_DIR', str(ROOT_DIR / '_banners'))
WP_USER = os.getenv('WP_USER')
WP_APP_PASSWORD = os.getenv('WP_APP_PASSWORD')

# ---------------------------------------------------------------------------
# HTTP helpers
# ---------------------------------------------------------------------------

def wp_get(path, params=None):
    if params is None:
        params = {}
    
    url = f"{API_BASE}{path}"
    headers = {
        'User-Agent': 'Jekyll-Importer/1.0',
        'Accept': 'application/json'
    }
    auth = (WP_USER, WP_APP_PASSWORD) if WP_USER and WP_APP_PASSWORD else None
    response = requests.get(url, params=params, headers=headers, auth=auth)
    if response.status_code != 200:
        raise Exception(f"HTTP {response.status_code} para {url}")
    
    body = response.json()
    total_pages = int(response.headers.get('X-WP-TotalPages', 1))
    
    return body, total_pages

def wp_get_all(path, params=None):
    if params is None:
        params = {}
    
    page = 1
    all_items = []
    
    try:
        while True:
            params.update({'page': page, 'per_page': 100})
            items, total_pages = wp_get(path, params)
            
            if not isinstance(items, list) or not items:
                break
                
            all_items.extend(items)
            if page >= total_pages:
                break
            page += 1
    except Exception as e:
        print(f"  AVISO: falha ao buscar {path} — {str(e)}")
        
    return all_items

# ---------------------------------------------------------------------------
# Taxonomy / embed helpers
# ---------------------------------------------------------------------------

def extract_terms(embedded, taxonomy):
    term_groups = embedded.get('wp:term', [])
    flattened = [item for sublist in term_groups for item in sublist]
    
    results = []
    for t in flattened:
        if isinstance(t, dict) and t.get('taxonomy') == taxonomy:
            name = html.unescape(str(t.get('name', ''))).strip()
            if name:
                results.append(name)
    return results

def first_term(embedded, taxonomy):
    terms = extract_terms(embedded, taxonomy)
    return terms[0] if terms else None

def featured_url(embedded):
    media = embedded.get('wp:featuredmedia')
    if not isinstance(media, list) or not media:
        return None
    
    first = media[0]
    if not isinstance(first, dict):
        return None
        
    return first.get('source_url')

def fetch_gallery(post_id, featured, acf_image_ids=None):
    try:
        if acf_image_ids:
            ids_str = ','.join(str(i) for i in acf_image_ids)
            media_items, _ = wp_get('/media', params={'include': ids_str, 'per_page': 100})
            id_to_url = {m['id']: m['source_url'] for m in media_items if m.get('source_url')}
            urls = [id_to_url[i] for i in acf_image_ids if i in id_to_url]
        else:
            media_items, _ = wp_get('/media', params={'parent': post_id, 'per_page': 100})
            urls = list(dict.fromkeys(m['source_url'] for m in media_items if m.get('source_url')))

        if featured:
            if featured in urls:
                urls.remove(featured)
            urls.insert(0, featured)

        return urls
    except Exception as e:
        print(f"  AVISO: galeria do post {post_id} — {str(e)}")
        return [featured] if featured else []

# ---------------------------------------------------------------------------
# File helpers
# ---------------------------------------------------------------------------

def format_preco(value):
    if value is None:
        return None
    try:
        v = float(value)
        formatted = f"{v:,.2f}"  # "72,700.00" US format
        return formatted.replace(',', 'X').replace('.', ',').replace('X', '.')  # "72.700,00"
    except (ValueError, TypeError):
        return value

def clean_wp_files(directory):
    dir_path = Path(directory)
    dir_path.mkdir(parents=True, exist_ok=True)
    
    for path in dir_path.glob('*.md'):
        try:
            content = path.read_text(encoding='utf-8')
            if re.search(r'^wp_id:', content, re.MULTILINE):
                path.unlink()
                print(f"    Removido: {path.name}")
            else:
                print(f"    Mantido (sem wp_id): {path.name}")
        except Exception as e:
            print(f"    ERRO ao ler {path.name}: {str(e)}")

def render_frontmatter(data):
    # PyYAML by default might format strings differently. 
    # Jekyll standard is simple. Using safe_dump.
    
    # Custom representer for None values to show them as empty keys
    def represent_none(self, _):
        return self.represent_scalar('tag:yaml.org,2002:null', '')

    yaml.add_representer(type(None), represent_none, Dumper=yaml.SafeDumper)
    
    # We want top-level fields to be ordered or at least consistent.
    # Ruby script used a manual builder to ensure specific formatting.
    # Let's try to match it closely.
    
    content = "---\n"
    # Filter out empty body if present in data
    body = data.pop('body', None)
    
    # Re-order logic to match Ruby's behavior if needed, 
    # but regular YAML is usually fine for Jekyll.
    yaml_str = yaml.safe_dump(data, allow_unicode=True, sort_keys=False, default_flow_style=False)
    content += yaml_str
    content += "---\n"
    
    if body:
        content += f"\n{body.strip()}\n"
    
    return content

def write_md(filepath, frontmatter, body=None):
    data = frontmatter.copy()
    if body:
        data['body'] = body
    
    content = render_frontmatter(data)
    Path(filepath).write_text(content, encoding='utf-8')

# ---------------------------------------------------------------------------
# Import: Carros
# ---------------------------------------------------------------------------

def import_carros():
    print("\n==> Importando carros...")
    
    carros = wp_get_all('/carro', params={'_embed': 1})
    total = len(carros)
    imported = 0
    seen_slugs = set()
    
    print(f"    {total} carro(s) encontrado(s) na API")
    
    for idx, carro in enumerate(carros):
        try:
            raw_slug = str(carro.get('slug', '')).strip()
            if not raw_slug:
                raw_slug = f"carro-{carro['id']}"
            
            slug = raw_slug
            if slug in seen_slugs:
                slug = f"{raw_slug}-{carro['id']}"
                print(f"  AVISO: slug duplicado '{raw_slug}', renomeado para '{slug}'")
            seen_slugs.add(slug)
            
            embedded = carro.get('_embedded', {})
            acf = carro.get('acf') or {}
            if isinstance(acf, list):
                acf = {}

            marca = first_term(embedded, 'marca')
            modelo = first_term(embedded, 'modelo')
            cambio = first_term(embedded, 'cambio')
            combustivel = first_term(embedded, 'combustivel')
            cor = first_term(embedded, 'cor')
            opcionais = extract_terms(embedded, 'opcional')
            
            anos = sorted(extract_terms(embedded, 'ano'))
            ano_str = "/".join(anos) if anos else None
            
            # Título do WordPress (prioridade)
            raw_title = html.unescape(carro.get('title', {}).get('rendered', '')).strip()
            
            # Título composto (fallback se rendered estiver vazio)
            parts = [p for p in [marca, modelo, ano_str] if p]
            composed = " ".join(parts)
            title = raw_title if raw_title else (composed if composed else slug)
            
            featured = featured_url(embedded)
            imagens = fetch_gallery(carro['id'], featured, acf.get('imagens'))
            
            body = html.unescape(carro.get('content', {}).get('rendered', '')).strip()
            
            acf_opcionais = acf.get('itens_e_opcionais')
            if isinstance(acf_opcionais, str) and acf_opcionais.strip():
                acf_opcionais = [o.strip() for o in acf_opcionais.replace('\n', ',').split(',') if o.strip()]
            elif not isinstance(acf_opcionais, list):
                acf_opcionais = []

            frontmatter = {
                'wp_id': carro['id'],
                'title': title if title else slug,
                'marca': marca,
                'modelo': modelo,
                'ano': ano_str,
                'km': acf.get('quilometragem'),
                'preco': format_preco(acf.get('preco')),
                'cambio': cambio,
                'combustivel': combustivel,
                'cor': cor,
                'portas': None,
                'destaque': None,
                'opcionais': opcionais if opcionais else acf_opcionais,
                'imagens': imagens,
            }
            
            filepath = Path(CARROS_DIR) / f"{slug}.md"
            write_md(filepath, frontmatter, body)
            
            imported += 1
            print(f"  [{idx + 1}/{total}] {slug}")
        except Exception as e:
            print(f"  ERRO ao importar carro ID {carro.get('id')}: {str(e)}")
            
    return imported

# ---------------------------------------------------------------------------
# Import: Banners
# ---------------------------------------------------------------------------

def import_banners():
    print("\n==> Importando banners...")
    
    banners = wp_get_all('/banner', params={'_embed': 1})
    total = len(banners)
    imported = 0
    
    print(f"    {total} banner(s) encontrado(s) na API")
    
    for idx, banner in enumerate(banners):
        try:
            slug = str(banner.get('slug', '')).strip()
            if not slug:
                slug = f"banner-{banner['id']}"
                
            embedded = banner.get('_embedded', {})
            imagem = featured_url(embedded)
            title = html.unescape(banner.get('title', {}).get('rendered', '')).strip()
            
            frontmatter = {
                'wp_id': banner['id'],
                'title': title if title else slug,
                'imagem': imagem,
                'link': None,
                'cta': None,
                'subtitulo': None,
                'ativo': True,
                'ordem': idx + 1,
            }
            
            filepath = Path(BANNERS_DIR) / f"{slug}.md"
            write_md(filepath, frontmatter)
            
            imported += 1
            print(f"  [{idx + 1}/{total}] {slug}")
        except Exception as e:
            print(f"  ERRO ao importar banner ID {banner.get('id')}: {str(e)}")
            
    return imported

# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def run_import():
    Path(CARROS_DIR).mkdir(parents=True, exist_ok=True)
    Path(BANNERS_DIR).mkdir(parents=True, exist_ok=True)
    
    print('==> Limpando arquivos gerados anteriormente...')
    print("    _carros/:")
    clean_wp_files(CARROS_DIR)
    print("    _banners/:")
    clean_wp_files(BANNERS_DIR)
    
    carros_importados = import_carros()
    banners_importados = import_banners()
    
    print("\n==> Concluído!")
    print(f"    Carros importados : {carros_importados}")
    print(f"    Banners importados: {banners_importados}")
    
    return {
        'carros': carros_importados,
        'banners': banners_importados
    }

if __name__ == "__main__":
    run_import()
    print("\nPróximo passo: bundle exec jekyll serve")
